lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+        """Returns dictionary with PV names as keys and values of PVs"""
+        return dict(zip(pvlist, epics.caget_many(pvlist)))
+
+    def evaluate_tao(self, tao, tao_cmds):
+        """Evaluate set of tao command, toggles lattice_calculation OFF/ON
+        in between command list"""
+        tao.cmd("set global lattice_calc_on = F")
+        tao.cmds(tao_cmds)
+        tao.cmd("set global lattice_calc_on = T")
+        output = self.get_output(tao)
+        return output
+
+    def kmod_to_bdes(self, tao, element, K=0):
+        """Returns BDES given K1 in the Bmad model"""
+        ele = tao.ele_gen_attribs(element)
+        bp = ele["E_TOT"] / 1e9 / 299.792458 * 1e4  # kG m
+        return ele["K1"] * bp * ele["L"]
+
+    def bdes_to_kmod(self, tao, element, bdes):
+        """Returns K1 given BDES"""
+        ele = tao.ele_gen_attribs(element)
+        bp = ele["E_TOT"] / 1e9 / 299.792458 * 1e4
+        # kG m
+        return bdes / ele["L"] / bp  # kG / m / kG m = 1/m^2
+
+    def plot_betas(self, output1, output2, **kwargs):
+        """Generates two figures with betas X and Y for two output runs"""
+        self.plot_beta_options = {
+            "title1": "",
+            "title2": "",
+            "label1": "Design",
+            "label2": "Model",
+            "figsize": (8, 4),
+        }
+        self.plot_beta_options.update(kwargs)
+        opt = self.plot_beta_options
+        fig1, ax1 = plt.subplots(figsize=(8, 4))
+        ax1.plot(
+            output1["ele.s"], output1["ele.a.beta"],
+            label=opt["label1"], linestyle="--"
+        )
+        ax1.plot(output2["ele.s"], output2["ele.a.beta"], label=opt["label2"])
+        plt.legend()
+        # Add energy to the rhs
+        ax12 = ax1.twinx()
+        ax12.plot(output2["ele.s"], output2["ele.e_tot"] / 1e9, color="red")
+        ax12.set_ylabel("Energy (GeV)")
+        efinal = output2["ele.e_tot"][-1] / 1e9
+        plt.title(f'{opt["title1"]} Final energy: {efinal:.2f} GeV')
+        ax1.set_xlabel("s (m)")
+        ax1.set_ylabel("Twiss Beta X (m)")
+        # itime = isotime()
+        fig2, ax2 = plt.subplots(figsize=(8, 4))
+        ax2.plot(
+            output1["ele.s"], output1["ele.b.beta"],
+            label=opt["label1"], linestyle="--"
+        )
+        ax2.plot(output2["ele.s"], output2["ele.b.beta"], label=opt["label2"])
+        plt.legend()
+        ax22 = ax2.twinx()
+        ax22.plot(output2["ele.s"], output2["ele.e_tot"] / 1e9, color="red")
+        ax22.set_ylabel("Energy (GeV)")
+        plt.title(f'{opt["title2"]} Final energy: {efinal:.2f} GeV')
+        ax2.set_xlabel("s (m)")
+        ax2.set_ylabel("Twiss Beta Y (m)")
+        axes_list = [ax1, ax12, ax2, ax22]
+        fig1.show()
+        fig2.show()
+        return fig1, fig2, axes_list
+
+    def get_output(self, tao):
+        """Returns dictionary of modeled parameters,
+        including element name, twiss and Rmats"""
+        # Output collecting
+        outkeys = [
```

Comment:
Will this ever change? maybe we save these in a yaml file that can be updated as needed vs. changing the code file? See the yaml files we make for [devices](https://github.com/slaclab/lcls-tools/blob/main/lcls_tools/common/devices/yaml/BC2.yaml). It can look different than these, but have a similar idea. Any static info that is not code could be put in a modeling yaml?

Reply:
I agree the yaml file is a good idea. There are a lot of opportunities to serialize hard coded data throughout this class, and a lot of opportunities to serialize functionality as well. We could reduce some of the is/else switchyard patterns if we're clear from the outset what beampath we're using, what devices are relevant, and so on.

However, the development I just described is a lot of work. If people will immediately find this tool useful, I suggest we push more or less as is, maybe with some light serialization. Then, we can create an issue for a more thorough OOP refactor of this class for the near future.

Reply:
Yes, I agree, would be good to use beam paths to reduce some of the if/else checks. Let's discuss what to push and what to write issues on in the meeting.